// challonge-node/dist/api/participants.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

var _api = new WeakMap();

var Participants = function () {
  function Participants(parent) {
    _classCallCheck(this, Participants);

    _api = parent;
  }

  _createClass(Participants, [{
    key: 'index',
    value: function index(tid) {
      var params = {};

      return new Promise(function (resolve, reject) {
        _api.request('GET', 'tournaments/' + tid + '/participants.json', params).then(function (response) {
          //remove 1st layer "participant"
          resolve(response.map(function (obj) {
            return obj.participant;
          }));
        }).catch(function (err) {
          reject(err.message);
        });
      });
    }
  }, {
    key: 'create',
    value: function create(tid, name) {
      var params = {
        name: name
      };

      return new Promise(function (resolve, reject) {
        _api.request('POST', 'tournaments/' + tid + '/participants.json', params).then(function (response) {
          resolve(response.participant);
        }).catch(function (err) {
          reject(err.message);
        });
      });
    }
  }, {
    key: 'show',
    value: function show(tid, pid) {
      var params = {};

      return new Promise(function (resolve, reject) {
        _api.request('GET', 'tournaments/' + tid + '/participants/' + pid + '.json', params).then(function (response) {
          resolve(response.participant);
        }).catch(function (err) {
          reject(err.message);
        });
      });
    }

    //http://api.challonge.com/v1/documents/participants/update

  }, {
    key: 'update',
    value: function update(tid, pid, params) {
      return new Promise(function (resolve, reject) {
        _api.request('PUT', 'tournaments/' + tid + '/participants/' + pid + '.json', params).then(function (response) {
          resolve(response.participant);
        }).catch(function (err) {
          reject(err.message);
        });
      });
    }
  }, {
    key: 'delete',
    value: function _delete(tid, pid) {
      var params = {};

      return new Promise(function (resolve, reject) {
        _api.request('DELETE', 'tournaments/' + tid + '/participants/' + pid + '.json', params).then(function (response) {
          resolve(response.participant);
        }).catch(function (err) {
          reject(err.message);
        });
      });
    }
  }]);

  return Participants;
}();

exports.default = Participants;
