// challonge-node/dist/challonge.js
'use strict';

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _requestPromiseNative = require('request-promise-native');

var _requestPromiseNative2 = _interopRequireDefault(_requestPromiseNative);

var _tournaments = require('./api/tournaments');

var _tournaments2 = _interopRequireDefault(_tournaments);

var _participants = require('./api/participants');

var _participants2 = _interopRequireDefault(_participants);

var _matches = require('./api/matches');

var _matches2 = _interopRequireDefault(_matches);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

var CHALLONGE_API_BASE_URI = "https://api.challonge.com/v1/";
var _api_key = new WeakMap();

var ChallongeAPI = function () {
  function ChallongeAPI(key) {
    _classCallCheck(this, ChallongeAPI);

    _api_key = key;
  }

  _createClass(ChallongeAPI, [{
    key: 'request',
    value: function request(method, uri, params) {
      params['api_key'] = _api_key;
      var options = {
        uri: CHALLONGE_API_BASE_URI + uri,
        method: method,
        qs: method === 'GET' || method === 'DELETE' || method === 'PUT' ? params : null,
        body: method === 'POST' ? params : null,
        json: true
      };

      return (0, _requestPromiseNative2.default)(options);
    }
  }]);

  return ChallongeAPI;
}();

exports.withAPIKey = function withAPIKey(key) {
  var instance = new ChallongeAPI(key);
  instance.tournaments = new _tournaments2.default(instance);
  instance.participants = new _participants2.default(instance);
  instance.matches = new _matches2.default(instance);
  return instance;
};
